spec/unit_test_pr_checker_spec.rb

```diff
@@ -0,0 +1,167 @@
+          allow(@dangerfile.git).to receive(:diff).and_return(git.diff)
+          allow(@my_plugin.github).to receive(:pr_labels).and_return(['unit-tests-exemption'])
+
+          @my_plugin.check_missing_tests
+
+          expect(@dangerfile.status_report[:errors]).to be_empty
+        end
+      end
+
+      it 'does not show that a PR with a custom bypass label is missing tests' do
+        changes_dict = {
+          'File1.java' => 'import java.utils.*;\n\n public class Abc { public static void main(String[] args) { println(''); } }',
+          'project/src/androidTest/java/org/test/ToolTest.kt' => 'class ToolTest { void testMethod() {} }',
+          'File2.kt' => 'class Abcdef(name: String) { public void testMe() { println(''); } }'
+        }
+
+        run_in_repo_with_diff(changes_dict: changes_dict) do |git|
+          bypass_label = 'ignore-no-tests'
+
+          allow(@dangerfile.git).to receive(:diff).and_return(git.diff)
+          allow(@my_plugin.github).to receive(:pr_labels).and_return([bypass_label])
+
+          @my_plugin.check_missing_tests(bypass_label: bypass_label)
+
+          expect(@dangerfile.status_report[:errors]).to be_empty
+        end
+      end
+
+      it 'does not show that a PR with custom classes / subclasses patterns are missing tests' do
+        exceptions = [
+          /ViewHelper$/
+        ].freeze
+
+        subclasses_exceptions = [
+          /BaseViewWrangler/
+        ].freeze
+
+        changes_dict = {
+          'File1.java' => 'import java.utils.*;\n\n public class Abc extends BaseViewWrangler { public static void main(String[] args) { println(''); } }',
+          'project/src/androidTest/java/org/test/ToolTest.kt' => 'class ToolTest { void testMethod() {} }',
+          'File2.kt' => 'class AbcdefViewHelper(name: String) { public void testMe() { println(''); } }'
+        }
+
+        run_in_repo_with_diff(changes_dict: changes_dict) do |git|
+          allow(@dangerfile.git).to receive(:diff).and_return(git.diff)
+
+          @my_plugin.check_missing_tests(classes_exceptions: exceptions, subclasses_exceptions: subclasses_exceptions)
+
+          expect(@dangerfile.status_report[:errors]).to be_empty
+        end
+      end
+    end
+
+    def run_in_repo_with_diff(changes_dict:)
+      Dir.mktmpdir do |dir|
+        Dir.chdir dir do
+          `git init -b master`
+
+          changes_dict.each do |key, _value|
+            file_path = "#{dir}/#{key}"
+            FileUtils.mkdir_p(File.dirname(file_path))
+            File.open(file_path, 'w') { |f| f.write "Initial #{key} content." }
+          end
+
+          `git add .`
+          `git commit -m "add files"`
+
+          # adds changes to the previously commited files to create a diff
+          changes_dict.each do |key, value|
+            File.open("#{dir}/#{key}", 'w') { |f| f.write(value) }
+          end
+
+          g = Git.open('.')
+          yield g
+        end
```

Comment:
While this should work, I'm unsure if it's a good thing to create a real git repo in a temp dir and do the real git actions (`git add`, `git commit`), as opposed to create a mock or test_double of the `Git` object that would return the mocked content.

Reply:
This was something I remember seeing in `release-toolkit` and thought would be cleaner/more readable than mocking diff strings; however, the test becomes a sort of integration test as well. I agree, I'll refactor it to use mocks and the tests as well, also to cover the other cases you mentioned in the other comment 👍